src/sinks/websocket_server/sink.rs

```diff
@@ -106,13 +112,36 @@ impl WebSocketListenerSink {
     async fn handle_connection(
         auth: Option<HttpServerAuthMatcher>,
         peers: Arc<Mutex<HashMap<SocketAddr, UnboundedSender<Message>>>>,
+        buffer: Arc<Mutex<VecDeque<(Uuid, Message)>>>,
         stream: MaybeTlsIncomingStream<TcpStream>,
         open_gauge: OpenGauge,
     ) -> Result<(), ()> {
+        // Base url for parsing request URLs that may be relative
+        let base_url = Url::parse("ws://localhost").ok();
         let addr = stream.peer_addr();
         debug!("Incoming TCP connection from: {}", addr);
 
+        let mut last_received = None;
+        let mut should_replay = false;
         let header_callback = |req: &Request, response: Response| {
+            let query_params = req.uri().query().unwrap_or("");
+            if query_params.contains("last_received") {
```

Comment:
Refactorig:
```suggestion
            if let Some(query_params) = req.uri().query().unwrap_or("") {
                if query_params.contains("last_received") {
```
Also, there is a lot of nesting here, it would read better in its own function

Reply:
I have moved some of these things into separate file, behind a trait. Let me know if that makes more sense.

Reply:
Looks much better 👍